framework/scripts/generate_test_list.py

```diff
@@ -74,31 +79,63 @@ def main():
+        if len(all_tests) != 0:
+            decls = [ f'extern struct test _test_{test};' for test in sorted(all_tests.keys()) ]
+            f.write('\n'.join(decls))
+            f.write('\n\n')
+
+        # per-gen list definition
+        if len(test_lists) != 0:
+            f.write('struct TestList {\n')
+            f.write('    uint64_t features;\n')
+            f.write('    const char* name;\n')
+            f.write('    const std::span<struct test* const> content;\n')
+            f.write('};\n\n')
+
+        # build all declared lists
         for name, tests in test_lists.items():
-            f.write(f'static constexpr struct test * const {name}_test_list[] = {{')
-            test_list = [ f'    &_test_{test}' for test in tests ]
-            f.write(',\n'.join(test_list))
-            f.write('\n};\n\n')
-        # write selector function
-        f.write('optional<const span<struct test * const>> get_test_list(const char *test_list_name) {\n')
-        if default_name is not None:
-            f.write(f'    if (!test_list_name) return {default_name}_test_list;\n')
+            f.write(f'static constexpr struct test * const {name}_test_list_[] = {{')
+            test_list = [ f'\n    &_test_{test}' for test in tests ]
+            f.write(','.join(test_list))
+            f.write('\n};\n')
+            f.write(f'static const struct TestList {name}_test_list = {{ {"cpu_" + name if name != "default" else 0}, "{name}", {name}_test_list_ }};\n\n')
+
+        # function to check if the list is selected, applied to non-default lists only
+        if other_names:
+            f.write('static bool test_list_matches(const TestList* test_list, const char* name)\n')
+            f.write('{\n')
+            f.write('#ifdef SANDSTONE\n')
+            f.write('    if ((name == nullptr) || (strcmp(name, "auto") == 0)) {\n')
+            f.write('        return (test_list->features & cpu_features) == test_list->features;\n')
+            f.write('    }\n')
+            f.write('#endif\n')
+            f.write('    return (name != nullptr) && (strcmp(name, test_list->name) == 0);\n')
+            f.write('}\n\n')
+
+        # selector function
+        f.write('std::optional<const std::span<struct test * const>> get_test_list(const char *name)\n{\n')
         for name in test_lists.keys():
-            f.write(f'    if  (!strcmp(test_list_name, "{name}")) return {name}_test_list;\n')
-        f.write('    return nullopt;\n')
+            if default_name is None or name != default_name:
+                f.write(f'    if (test_list_matches(&{name}_test_list, name)) return {name}_test_list.content;\n')
+        if default_name is not None:
+            f.write('    if (name && (strcmp(name, "default") != 0)) {\n')
+            f.write('        logging_printf(LOG_LEVEL_QUIET, "# ERROR: list for %s not found. Using default\\n", name);\n')
+            f.write('    }\n')
+            f.write(f'    return {default_name}_test_list.content;\n')
```

Comment:
I may need in the future to have this return the `TestList` object so we can include the name of the list in the log output.

Reply:
Appropriate function added (with a bit confusing name, but I'd prefer keep modification in the generator only). "C" struct definition moved to generated header (I wonder if it should be hpp with using, as it won't compile with "C").
The structure contains source of the list to improve the experience.